### ISSUE TITLE
Vertaisarviointi

### ISSUE THREAD

**Issue by PlatinumFoxTail** [carlotta-mah/keystroke_gen#5]
### Evaluation based cloned repository from main branch 14th of Feb 2024 @ 15:51

**General evaluation incl. documentation folder and other files:**

- In the specifications document it would be helpful to open up a bit what biomteric keystroke data is. I had to read and google a bit, before this fascinating area opened up to me :)
- It could be an idea to document how to set up the project for other e.g. in a manual document. Even though not a major thing, it took a few minutes to look through the files and figure out to run pip install -r ./requirements.txt in the terminal and run neural_network.py.
- Is the trained neural network in neural_network.py, utilized for some type of predictions somewhere? I was not able to find some kind of user-interface, where the trained model would be utilized. Maybe this is still coming?

**Model evaluation i.e. neural_network.py and data_reader.py:**

- It required a bit of searching in neural_network.py file to identify where to insert the hyperparameters i.e. neurons in each layer, epochs, and learning rate. Maybe that is coming up in somekind of user-interface or alternatively to be commented in the code?
- One thing that I did not quite get is why having several options for the _activation_function method in the neural_network.py? I guess it gives more optionality, but did not get why that is necessary. Curious to learn about that more :)
- As what it comes to the neural network model in the neural_network.py I think the structure is quite clear and the naming of methods and variables gives a good description.
- I also reviewed quckly the data_read.py file, and it seemed OK as well.

**Test evaluation i.e. data_reader_test.py and neural_network_test.py:**

- It took some time for me to get the tests running and I am not sure I was able to run them as intended (probably not). It might be an idea to write e.g. in a test document how to run the tests. Here an explanation how I run the tests, so it gives a context to the test evaluation afterwards in this review:
    • In the test folder in the terminal: python3 -m venv venv
    • In the test folder in the terminal: source venv/bin/activate
    • In the root folder in the terminal: pip install -r requirements.txt
    • Changed src.datareader → datareader and src.neural_network → neural network in the test files
    • Moved files in src folder to test folder
    • In the root folder in the terminal → pytest

- When running the test as above, I get 40% coverage of data_reader.py and 100% coverage of neural_network.py. So that seems to be on good track. Further on all the 5 tests are passes, which is also great.
- In the test_neural_network test it seems feasible to assert that the shapes of the predictions are of same shape as the test and training data. Maybe the plot could have some lables so it becomes immediately clear that just plotting the generated 1000 observations?
- The test_forward_propagation() test seems feasible to check that right number of neurons remains.
- The test_backward_propagation seems also viable to test that the weights and biases are updated. Could that be further refined to check that they are updated into correct direction? Just an idea that came up.
- There seems to be more tests coming up for checking data_reader.py, but the current tests in data_reader_test.py seems feasible. Based on the test_data.txt, it seems to be that there are quite many things that can be missing or wrong in the input data. If I understood that right, then maybe one idea would be to list what are the most likely severe error in the input data and start building the tests for those in the data_reader_test.py? You have probably thought about this already, but thought I’ll share it just to assure you are not overwhelmed with the many things that could be tested :)
- If you are interested in adding automated testing and coverage here is some intro on that starting from “9. GitHub Actions, osa 1” (material is in Finnish, but there are some nice illustrative pictures as well): https://ohjelmistotuotanto-hy.github.io/tehtavat1
